compiler/src/dotty/tools/dotc/transform/sjs/PrepJSInterop.scala

```diff
@@ -0,0 +1,1180 @@
+import dotty.tools.backend.sjs.JSDefinitions.jsdefn
+
+/** A macro transform that runs after typer and before pickler to perform
+ *  additional Scala.js-specific checks and transformations necessary for
+ *  interoperability with JavaScript.
+ *
+ *  It performs the following functions:
+ *
+ *  - Sanity checks for the js.Any hierarchy
+ *  - Annotate subclasses of js.Any to be treated specially
+ *  - Create JSExport methods: Dummy methods that are propagated
+ *    through the whole compiler chain to mark exports. This allows
+ *    exports to have the same semantics than methods.
+ *
+ *  This is the equivalent of `PrepJSInterop` in Scala 2, minus the handling
+ *  of `scala.Enumeration`.
+ *
+ *  The reason for making this a macro transform is that some functions (in particular
+ *  all the checks that behave differently depending on properties of classes in
+ *  the enclosing class chain) are naturally top-down and don't lend themselves to the
+ *  bottom-up approach of a mini phase.
+ *
+ *  In addition, the addition of export forwarders must be done before pickling to
+ *  be signature-compatible with scalac, and there are only macro transforms before
+ *  pickling.
+ */
+class PrepJSInterop extends MacroTransform with IdentityDenotTransformer { thisPhase =>
+  import PrepJSInterop._
+  import tpd._
+
+  override def phaseName: String = PrepJSInterop.name
+
+  override def isEnabled(using Context): Boolean =
+    ctx.settings.scalajs.value
+
+  override def changesMembers: Boolean = true // the phase adds export forwarders
+
+  override def runOn(units: List[CompilationUnit])(using Context): List[CompilationUnit] = {
+    sjsPlatform.perRunInfo.clear()
+    super.runOn(units)
+  }
+
+  protected def newTransformer(using Context): Transformer =
+    new ScalaJSPrepJSInteropTransformer
+
+  class ScalaJSPrepJSInteropTransformer extends Transformer with Checking {
+    import PrepJSExports._
+
+    /** Kind of the directly enclosing (most nested) owner. */
+    private var enclosingOwner: OwnerKind = OwnerKind.None
+
+    /** Cumulative kinds of all enclosing owners. */
+    private var allEnclosingOwners: OwnerKind = OwnerKind.None
+
+    /** Nicer syntax for `allEnclosingOwners is kind`. */
+    private def anyEnclosingOwner: OwnerKind = allEnclosingOwners
+
+    private def enterOwner[A](kind: OwnerKind)(body: => A): A = {
+      require(kind.isBaseKind, kind)
+      val oldEnclosingOwner = enclosingOwner
+      val oldAllEnclosingOwners = allEnclosingOwners
+      enclosingOwner = kind
+      allEnclosingOwners |= kind
+      try {
+        body
+      } finally {
+        enclosingOwner = oldEnclosingOwner
+        allEnclosingOwners = oldAllEnclosingOwners
+      }
+    }
+
+    /** Whether to check that we have proper literals in some crucial places.
+     *
+     *  This is always true in dotc. We keep the definition so that the code
+     *  code can be as similar as possible to the scalac phase.
```

Comment:
This reasoning (keep things close together) seems to be inconsistently applied across this file. For example: `fixPublicBeforeTyper` has been fully removed (instead of replaced by an identity function).

I think we should stick to one approach here. Personally, I think it is probably worth diverging, but doing it properly (i.e. remove things we don't need, etc.).

OTOH, we should invest to share tests suites as much as we can to avoid (and/or identify) feature/behavior divergence.